File: src/main/ruby/groovyclient.rb
#!/usr/bin/ruby
#
# Copyright 2009-2013 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
require 'socket'
require 'base64'

#-------------------------------------------
# Constants
#-------------------------------------------

DESTHOST = "127.0.0.1" # for security
DESTPORT = ENV.fetch("GROOVYSERVER_PORT", 1961)
IS_WINDOWS = RUBY_PLATFORM.downcase =~ /mswin(?!ce)|mingw|cygwin|bccwin/
HOME_DIR = IS_WINDOWS ? ENV['USERPROFILE'] : ENV['HOME']
AUTHTOKEN_FILE_BASE = HOME_DIR + "/.groovy/groovyserv/authtoken"
GROOVYSERVER_CMD = File.expand_path(ENV.fetch("GROOVYSERV_HOME", File.dirname($0)+"/..") + "/bin/groovyserver")
VERSION_MESSAGE = "GroovyServ Version: Client: @GROOVYSERV_VERSION@ (.ruby)"

ERROR_INVALID_AUTHTOKEN = 201
ERROR_CLIENT_NOT_ALLOWED = 202

#-------------------------------------------
# Classes
#-------------------------------------------

class Options
  attr_reader :client, :server
  def initialize
    @client = {
      :without_invoking_server => false,
      :host => DESTHOST,
      :port => DESTPORT,
      :authtoken => nil,
      :quiet => false,
      :env_all => false,
      :env_include_mask => [],
      :env_exclude_mask => [],
      :help => false,
      :version => false,
      :groovyserver_opt => [],
    }
    @server = {
      :help => false,
      :version => false,
      :args => [],
    }
  end
  def need_to_invoke_server?
    return false if @client[:without_invoking_server]
    @client[:groovyserver_opt].any? {|v| ['-k', '-r'].include?(v) }
  end
end

#-------------------------------------------
# Global variables
#-------------------------------------------

$options = nil    # FIXME shouldn't use global variables

#-------------------------------------------
# Functions
#-------------------------------------------

def usage()
  puts "\
usage: groovyclient.rb -C[option for groovyclient] [args/options for groovy]
options:
  -Ch,-Chelp                       show this usage
  -Cs,-Chost                       specify the host to connect to groovyserver
  -Cp,-Cport <port>                specify the port to connect to groovyserver
  -Ca,-Cauthtoken <authtoken>      specify the authtoken
  -Ck,-Ckill-server                kill the running groovyserver
  -Cr,-Crestart-server             restart the running groovyserver
  -Cq,-Cquiet                      suppress statring messages
  -Cenv <substr>                   pass environment variables of which a name
                                   includes specified substr
  -Cenv-all                        pass all environment variables
  -Cenv-exclude <substr>           don't pass environment variables of which a
                                   name includes specified substr
  -Cv,-Cversion                    display the GroovyServ version"
end

def start_server(args)
  unless FileTest.executable? GROOVYSERVER_CMD
    STDERR.puts "ERROR: Command not found: #{GROOVYSERVER_CMD}"
    exit 1
  end
  if $options.client[:quiet]
    args << "-q"
  else
    command_str = "'#{GROOVYSERVER_CMD}' -p #{$options.client[:port]} #{args.join(' ')}"
    STDERR.printf "Invoking server: %s\n", command_str
  end
  system(GROOVYSERVER_CMD, "-p", $options.client[:port].to_s, *args)
end

def session(socket, args)
  send_command(socket, args)
  while true do
    IO.select([socket, $stdin]).each do |ins|
      if ins[0] == socket
        handle_socket(socket)
      elsif ins[0] == $stdin
        handle_stdin(socket)
      end
    end
  end
end

def send_envvars(socket)
  ENV.each{|key,value|
    if $options.client[:env_all] || $options.client[:env_include_mask].any?{|item| key.include?(item) }
      if !$options.client[:env_exclude_mask].any?{|item| key.include?(item) }
        socket.puts "Env: #{key}=#{value}"
      end
    end
  }
end

def send_command(socket, args)
  socket.puts "Cwd: #{current_dir}"
  args.each do |arg|
    # why using gsub? It's because Base64.encode64 happens to break lines automatically.
    # TODO using default encoding.
    socket.puts "Arg: #{Base64.encode64(arg).gsub(/\s/, '')}"
  end
  socket.puts "AuthToken: #{authtoken}"
  send_envvars(socket)
  if ENV['CLASSPATH']
    socket.puts "Cp: #{ENV['CLASSPATH']}"
  end
  socket.puts ""
end

def authtoken
  $options.client[:authtoken] || File.open(AUTHTOKEN_FILE_BASE + "-" + $options.client[:port].to_s) { |f| f.read }
end

def current_dir()
  if IS_WINDOWS
    # native path expression including a drive letter is needed.
    `cmd.exe /c cd`.strip # FIXME it's ugly...
  else
    Dir::pwd
  end
end

def handle_socket(socket)
  headers = read_headers(socket)
  if headers.empty?
    exit 1
  end
  if headers['Status']
    if $options.server[:help]
      puts "\n"
      usage()
    end
    if $options.server[:version]
      puts VERSION_MESSAGE
    end
    status_code = headers['Status'].to_i
    if status_code == ERROR_INVALID_AUTHTOKEN
      STDERR.puts "ERROR: rejected by groovyserv because of invalid authtoken"
    elsif status_code == ERROR_CLIENT_NOT_ALLOWED
      STDERR.puts "ERROR: rejected by groovyserv because of not allowed client address"
    end
    exit status_code
  end
  data = socket.read(headers['Size'].to_i)
  return unless data

  if headers['Channel'] == 'out'
    $stdout.print data
    $stdout.flush
  elsif headers['Channel'] == 'err'
    $stderr.print data
    $stderr.flush
  end
end

$closedStdin = false # FIXME
def handle_stdin(socket)
  return if $closedStdin
  begin
    data = $stdin.read_nonblock(512)
  rescue EOFError
    socket.write "Size: 0\n\n"
    $closedStdin = true
  else
    socket.write "Size: #{data.length}\n\n"
    socket.write data
  end
end

def send_interrupt(socket)
    socket.write "Size: -1\n\n"
    socket.close()
end

def read_headers(socket)
  headers = {}
  while (line = socket.gets) != nil && line != "\n" do
    line.chomp!
    /([a-zA-Z]+): (.+)/ =~ line
    headers[$1] = $2
  end
  headers
end

def parse_option(args)
  options = Options.new
  args.each_with_index do |arg, i|
    case arg
    when "-Cwithout-invoking-server"
      options.client[:without_invoking_server] = true
    when "-Ch", "-Chost"
      options.client[:host] = args.delete_at(i + 1)
    when "-Cp", "-Cport"
      port = args.delete_at(i + 1)
      unless port =~ /^[0-9]+$/
        raise "Invalid port number #{port} for #{arg}"
      end
      options.client[:port] = port
    when "-Ca", "-Cauthtoken"
      options.client[:authtoken] = args.delete_at(i + 1)
    when "-Ck" , "-Ckill-server"
      options.client[:groovyserver_opt] << "-k"
    when "-Cr", "-Crestart-server"
      options.client[:groovyserver_opt] << "-r"
    when "-Cq", "-Cquiet"
      options.client[:groovyserver_opt] << "-q"
      options.client[:quiet] = true
    when "-Cenv-all"
      options.client[:env_all] = true
    when "-Cenv"
      val = args.delete_at(i + 1)
      unless val
        raise "Invalid mask string #{val} for #{arg}"
      end
      options.client[:env_include_mask] << val
    when "-Cenv-exclude"
      val = args.delete_at(i + 1)
      unless val
        raise "Invalid mask string #{val} for #{arg}"
      end
      options.client[:env_exclude_mask] << val
    when "-Ch", "-Chelp"
      options.client[:help] = true
    when "--help", "-help", "-h"
      options.server[:help] = true
      options.server[:args] << arg
    when "-Cv", "-Cversion"
      options.client[:version] = true
    when "--version", /^-v/
      options.server[:version] = true
      options.server[:args] << arg
    when /-C.*/
      raise "Unknown option #{arg}"
    else
      options.server[:args] << arg
    end
  end
  if options.server[:args].empty?
    # display additionally client's usage at the end of session
    options.server[:help] = true
  end
  options
end

#-------------------------------------------
# Main
#-------------------------------------------

# Parsing options
begin
  $options = parse_option(ARGV)
rescue => e
  STDERR.puts "ERROR: #{e.message}"
  usage()
  exit 1
end
#puts "Original ARGV: #{ARGV.inspect}"
#puts "Parsed options: #{$options.inspect}"

# Only show usage (highest priority)
if $options.client[:help]
  usage()
  exit 0
end

# Only show version (highest priority)
if $options.client[:version]
    puts VERSION_MESSAGE
  exit 0
end

# Start or stop server when specified
if $options.need_to_invoke_server?
  start_server($options.client[:groovyserver_opt].uniq)
  if $options.client[:groovyserver_opt].include?("-k")
    exit 0
  end
end

# Invoke script (before, start server if down)
failCount = 0
begin
  TCPSocket.open($options.client[:host], $options.client[:port]) { |socket|
    Signal.trap(:INT) {
      send_interrupt(socket)
      exit 8
    }
    session(socket, $options.server[:args])
  }
rescue Errno::ECONNREFUSED
  if $options.client[:without_invoking_server]
    STDERR.puts "ERROR: groovyserver isn't running"
    exit 9
  end
  if failCount >= 3
    STDERR.puts "ERROR: Failed to start up groovyserver: #{GROOVYSERVER_CMD}"
    exit 1
  end
  start_server([])
  sleep 3
  failCount += 1
  retry
rescue Errno::ECONNRESET, Errno::EPIPE
  # normally exit if reset by peer or broken pipe
  exit 0
rescue => e
  STDERR.puts "ERROR: #{e.message}"
  STDERR.puts e.backtrace
  exit 1
end

File: src/main/go/groovyclient.go
/*
 * Copyright 2014 the original author or authors.
 *
 * Licensed under the Apache License, version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package main

import (
	gs "./common"
	"bufio"
	b64 "encoding/base64"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net"
	"os"
	"os/exec"
	"os/signal"
	"path"
	"regexp"
	"strconv"
	"strings"
	"syscall"
)

const (
	DEFAULT_HOST             = "localhost"
	DEFAULT_PORT             = 1961
	ERROR_INVALID_AUTHTOKEN  = 201
	ERROR_CLIENT_NOT_ALLOWED = 202
	BUFFER_SIZE              = 4096
	USAGE                    = `usage: groovyclient -C[option for groovyclient] [args/options for groovy]
options:
  -Ch,-Chelp                       show this usage
  -Cs,-Chost                       specify the host to connect to groovyserver
  -Cp,-Cport <port>                specify the port to connect to groovyserver
  -Ca,-Cauthtoken <authtoken>      specify the authtoken
  -Ck,-Ckill-server                kill the running groovyserver
  -Cr,-Crestart-server             restart the running groovyserver
  -Cq,-Cquiet                      suppress statring messages
  -Cenv <substr>                   pass environment variables of which a name
                                   includes specified substr
  -Cenv-all                        pass all environment variables
  -Cenv-exclude <substr>           don't pass environment variables of which a
                                   name includes specified substr
  -Cv,-Cversion                    display the GroovyServ version
  -Cdebug                          display console log`
)

var (
	HOME_DIR            = gs.HomeDir()
	GROOVYSERV_HOME_DIR = gs.Env("GROOVYSERV_HOME", path.Join(path.Dir(os.Args[0]), ".."))
	VERSION_MESSAGE     = "GroovyServ Version: Client: " + GROOVYSERV_VERSION
	GROOVYSERV_VERSION  = "x.x" // replaced by ldflags
)

type (
	ClientOptions struct {
		host            string
		port            int
		authToken       string
		quiet           bool
		envAll          bool
		envIncludeMask  []string
		envExcludeMask  []string
		help            bool
		version         bool
		debug           bool
		startServerOpts []string
	}

	ServerOptions struct {
		help    bool
		version bool
		args    []string
	}

	Options struct {
		client ClientOptions
		server ServerOptions
	}
)

func NewOptions() Options {
	return Options{
		client: ClientOptions{
			host:            gs.Env("GROOVYSERV_HOST", DEFAULT_HOST),
			port:            gs.EnvInt("GROOVYSERV_PORT", DEFAULT_PORT),
			authToken:       "",
			quiet:           false,
			envAll:          false,
			envIncludeMask:  []string{},
			envExcludeMask:  []string{},
			help:            false,
			version:         false,
			debug:           false,
			startServerOpts: []string{},
		},
		server: ServerOptions{
			help:    false,
			version: false,
			args:    []string{},
		},
	}
}

func ParseOptions(args gs.Args) Options {
	opts := NewOptions()
	for i := 1; i < len(args); i++ {
		arg, param := args[i], ""

		// Replacing "-v.*" -> "--version". This behavior is as same as regular groovy command.
		if regexp.MustCompile(`^-v.*`).MatchString(arg) {
			arg = "-v"
		}

		switch arg {
		case "-Cs", "-Chost":
			i, param = args.Param(i, arg)
			opts.client.host = param
		case "-Cp", "-Cport":
			i, param = args.Param(i, arg)
			port, err := strconv.Atoi(param)
			if err != nil {
				panic(fmt.Sprintf("could not parse port number '%s'", param))
			}
			opts.client.port = port
		case "-Ca", "-Cauthtoken":
			i, param = args.Param(i, arg)
			opts.client.authToken = param
		case "-Cq", "-Cquiet":
			opts.client.quiet = true
		case "-Cenv-all":
			opts.client.envAll = true
		case "-Cenv", "-Cenv-exclude":
			i, param = args.Param(i, arg)
			if arg == "-Cenv" {
				opts.client.envIncludeMask = append(opts.client.envIncludeMask, param)
			} else {
				opts.client.envExcludeMask = append(opts.client.envExcludeMask, param)
			}
		case "-Ch", "-Chelp":
			opts.client.help = true
		case "-h", "--help", "-help":
			opts.server.help = true
			opts.server.args = append(opts.server.args, arg)
		case "-Cv", "-Cversion":
			opts.client.version = true
		case "--version", "-v":
			opts.server.version = true
			opts.server.args = append(opts.server.args, arg)
		case "-Cdebug":
			opts.client.debug = true
		case "-Ck", "-Ckill-server":
			opts.client.startServerOpts = append(opts.client.startServerOpts, "-k")
		case "-Cr", "-Crestart-server":
			opts.client.startServerOpts = append(opts.client.startServerOpts, "-r")
		default:
			if m, _ := regexp.MatchString("-C.*", arg); m {
				panic(fmt.Sprintf("unrecognized option %s", arg))
			}
			opts.server.args = append(opts.server.args, arg)
		}
	}
	// To show a help usage if no args for server.
	if len(opts.server.args) == 0 {
		opts.server.help = true
	}
	return opts
}

func HandleSignal(writer io.Writer) {
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	for sig := range c {
		log.Printf("Signaled: %#v", sig)
		gs.Write(writer, "Cmd: interrupt\n\n")
		os.Exit(1)
	}
}

func StartSession(opts Options) {
	// Connecting to server process.
	conn, err := gs.Connect(opts.client.host, opts.client.port)
	if err != nil {
		// If a server isn't running, try to start it once.
		if oerr, ok := err.(*net.OpError); ok && syscall.ECONNREFUSED == oerr.Err {
			StartServer(opts)
			conn, err = gs.Connect(opts.client.host, opts.client.port)
			if err != nil {
				panic(fmt.Sprintf("could not connect to server: %s", err.Error()))
			}
		} else {
			panic(fmt.Sprintf("could not connect to server: %s", err.Error()))
		}
	}
	defer conn.Close()

	go HandleSignal(conn)
	WriteInvocationRequest(conn, opts)
	go WriteStreamRequest(conn)
	ReadResponse(conn, opts)
}

func WriteStreamRequest(writer io.Writer) {
	reader := bufio.NewReader(os.Stdin)
	for {
		input, err := reader.ReadString('\n')
		if err != nil {
			gs.Write(writer, "Size: 0\n\n")
			break
		}
		gs.Write(writer, fmt.Sprintf("Size: %d\n\n", len(input)))
		gs.Write(writer, input) // without LF
	}
}

func ReadResponse(reader io.Reader, opts Options) {
	rd := bufio.NewReaderSize(reader, BUFFER_SIZE)
	buff := make([]byte, BUFFER_SIZE)
	for {
		// Reading headers
		headers := ReadHeaders(rd)
		if len(headers) == 0 {
			panic(fmt.Sprintf("no header"))
		}

		// Status:
		statusCodeStr, ok := headers.Value("Status")
		if ok {
			log.Println("Status found:", statusCodeStr)

			if opts.server.help {
				fmt.Println("")
				fmt.Println(USAGE)
			}
			if opts.server.version {
				fmt.Println(VERSION_MESSAGE)
			}

			statusCode, _ := strconv.Atoi(statusCodeStr)
			switch statusCode {
			case ERROR_INVALID_AUTHTOKEN:
				panic(fmt.Sprintf("invalid authtoken"))
			case ERROR_CLIENT_NOT_ALLOWED:
				panic(fmt.Sprintf("client address not allowed %s:%d", opts.client.host, opts.client.port))
			}

			os.Exit(statusCode)
		}

		// Size:
		bodySizeStr, ok := headers.Value("Size")
		if !ok {
			panic(fmt.Sprintf("no response: %#v", headers))
		}
		bodySize, _ := strconv.Atoi(bodySizeStr)

		// Channel:
		channel, _ := headers.Value("Channel")
		targetStream := os.Stdout
		if channel == "err" {
			targetStream = os.Stderr
		}

		// Reading body
		var err error
		for n, read := 0, 0; read < bodySize; read += n {
			n, err = rd.Read(buff[0:bodySize])
			if err == io.EOF {
				break
			}
			if err != nil {
				panic(fmt.Sprintf("could not read body from server: %s", err.Error()))
			}
			fmt.Fprint(targetStream, string(buff[0:n]))
		}
	}
}

type Headers map[string]string

func (headers Headers) Value(key string) (value string, ok bool) {
	value, ok = headers[key]
	return strings.Trim(value, " "), ok
}

func ReadHeaders(reader *bufio.Reader) Headers {
	headers := Headers{}
	for {
		line, err := gs.ReadLine(reader)
		if err != io.EOF && err != nil { // EOF isn't error
			panic(fmt.Sprintf("could not read headers from server: %s", err.Error()))
		}
		tokens := strings.SplitN(line, ":", 2)
		if len(tokens) < 2 {
			return headers
		}
		key := tokens[0]
		value := tokens[1]
		headers[key] = value
		log.Printf("Key/Value: '%s' => '%s'\n", key, value)
	}
	return headers
}

func WriteInvocationRequest(writer io.Writer, opts Options) {
	gs.WriteLine(writer, "Cwd: /tmp")
	gs.WriteLine(writer, "Auth: "+AuthToken(opts.client.port, opts))
	for _, arg := range opts.server.args {
		gs.WriteLine(writer, "Arg: "+b64.StdEncoding.EncodeToString([]byte(arg)))
	}
	WriteEnv(writer, opts)
	if cp := gs.Env("CLASSPATH", ""); cp != "" {
		gs.WriteLine(writer, "Cp: "+cp)
	}
	gs.WriteLine(writer, "")
}

func WriteEnv(writer io.Writer, opts Options) {
	for _, env := range os.Environ() {
		name := strings.SplitN(env, "=", 2)[0]
		maskMatcher := func(mask string) bool {
			return strings.Index(name, mask) >= 0
		}
		if opts.client.envAll || gs.ArrayFuncAny(opts.client.envIncludeMask, maskMatcher) {
			if !gs.ArrayFuncAny(opts.client.envExcludeMask, maskMatcher) {
				gs.WriteLine(writer, "Env: "+env)
			}
		}
	}
}

func AuthToken(port int, opts Options) string {
	// User specified token has the highest priority.
	if opts.client.authToken != "" {
		return opts.client.authToken
	}

	// Reading a stored token from a file.
	authTokenPath := path.Join(HOME_DIR, fmt.Sprintf(".groovy/groovyserv/authtoken-%d", port))
	body, err := ioutil.ReadFile(authTokenPath)
	if err != nil {
		panic(fmt.Sprintf("could not read authtoken file: %s", err.Error()))
	}
	authToken := string(body)
	log.Println("AuthToken:", authToken)
	return authToken
}

func StartServer(opts Options) {
	serverCmdFile := path.Join(GROOVYSERV_HOME_DIR, "bin/groovyserver")
	if !gs.FileExists(serverCmdFile) {
		panic(fmt.Sprintf("server command not found %s", serverCmdFile))
	}
	log.Printf("groovyserver command path: %s", serverCmdFile)

	cmd := exec.Command(serverCmdFile, "-p", fmt.Sprint(opts.client.port), strings.Join(opts.client.startServerOpts, " "))
	if !opts.client.quiet {
		fmt.Fprintf(os.Stderr, "Start server: %s -p %d %s\n", serverCmdFile, opts.client.port, strings.Join(opts.client.startServerOpts, " "))
		cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
	}
	err := cmd.Run()
	if err != nil {
		panic(fmt.Sprintf("could not start groovyserver: %s", err.Error()))
	}
}

func main() {
	defer gs.ExitIfPanic()

	// Parsing options
	opts := ParseOptions(os.Args)

	// Setting log level
	if !opts.client.debug {
		log.SetOutput(ioutil.Discard)
	}
	log.Println("Original arguments:", os.Args)
	log.Printf("Parsed options: %#v", opts) // must be after affecting -Cdebug option

	// Only show usage (highest priority)
	if opts.client.help {
		fmt.Println(USAGE)
		os.Exit(0)
	}

	// Only show version (highest priority)
	if opts.client.version {
		fmt.Println(VERSION_MESSAGE)
		os.Exit(0)
	}

	// Start or kill server when specified
	if gs.ArrayAny(opts.client.startServerOpts, "-k", "-r") {
		StartServer(opts)
		if gs.ArrayAny(opts.client.startServerOpts, "-k") {
			os.Exit(0)
		}
	}

	// Make a groovy script evaluate on a server
	StartSession(opts)

	// Only show version (highest priority)
	if opts.server.version {
		fmt.Println(VERSION_MESSAGE)
		os.Exit(0)
	}
}
